File: android/src/main/kotlin/com/example/multi_ads/pangle_global/PangleAdsHandler.kt
package com.example.multi_ads.pangle_global

import android.app.Activity
import android.content.Context
import android.view.View
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.bytedance.sdk.openadsdk.api.init.PAGConfig
import com.bytedance.sdk.openadsdk.api.init.PAGSdk
import com.bytedance.sdk.openadsdk.api.nativeAd.PAGNativeAd

class PangleAdsHandler(
    private val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
) {
    private val channel: MethodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "multi_ads/pangle_global")
    private val splashEventChannel: EventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "multi_ads/pangle_global/splash_events")
    private val interstitialEventChannel: EventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "multi_ads/pangle_global/interstitial_events")
    
    private val context: Context = flutterPluginBinding.applicationContext
    private var activity: Activity? = null
    
    private val nativeEventChannel: EventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "multi_ads/pangle_global/native_events")

    private var splashEventSink: EventChannel.EventSink? = null
    private var interstitialEventSink: EventChannel.EventSink? = null
    private var nativeEventSink: EventChannel.EventSink? = null
    
    private var splashAdManager: SplashAdManager? = null
    private var interstitialAdManager: InterstitialAdManager? = null
    private val bannerAdManagers = mutableMapOf<String, BannerAdManager>()
    private val nativeAdManagers = mutableMapOf<String, NativeAdManager>()

    init {
        channel.setMethodCallHandler { call, result ->
            onMethodCall(call, result)
        }
        
        splashEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                splashEventSink = events
            }

            override fun onCancel(arguments: Any?) {
                splashEventSink = null
            }
        })
        
        interstitialEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                interstitialEventSink = events
            }

            override fun onCancel(arguments: Any?) {
                interstitialEventSink = null
            }
        })
        
        nativeEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                nativeEventSink = events
                for (mgr in nativeAdManagers.values) {
                    mgr.updateEventSink(events)
                }
            }

            override fun onCancel(arguments: Any?) {
                nativeEventSink = null
                for (mgr in nativeAdManagers.values) {
                    mgr.updateEventSink(null)
                }
            }
        })
        
        // Register platform views
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "multi_ads/pangle_global/banner",
            BannerAdViewFactory(flutterPluginBinding.binaryMessenger)
        )
        
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "multi_ads/pangle_global/banner_container",
            BannerAdContainerFactory(this)
        )
        
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "multi_ads/pangle_global/native",
            NativeAdViewFactory(this)
        )
    }
    
    fun setActivity(activity: Activity?) {
        this.activity = activity
    }
    
    fun getBannerView(listenerId: String): View? {
        return bannerAdManagers[listenerId]?.getBannerView()
    }
    
    fun getNativeAd(listenerId: String): PAGNativeAd? {
        return nativeAdManagers[listenerId]?.nativeAd
    }
    
    fun getNativeEventSink(): EventChannel.EventSink? {
        return nativeEventSink
    }

    fun sendNativeSwipe(listenerId: String, direction: Int) {
        nativeEventSink?.success(mapOf(
            "listenerId" to listenerId,
            "event" to "onAdSwipe",
            "direction" to direction
        ))
    }

    private fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                val appId = call.argument<String>("appId") ?: ""
                val debug = call.argument<Boolean>("debug") ?: false
                initialize(appId, debug, result)
            }
            "loadSplashAd" -> {
                val slotId = call.argument<String>("slotId") ?: ""
                val timeout = call.argument<Int>("timeout") ?: 3000
                loadSplashAd(slotId, timeout, result)
            }
            "loadInterstitialAd" -> {
                val slotId = call.argument<String>("slotId") ?: ""
                loadInterstitialAd(slotId, result)
            }
            "showInterstitialAd" -> {
                showInterstitialAd(result)
            }
            "loadBannerAd" -> {
                val slotId = call.argument<String>("slotId") ?: ""
                val adSize = call.argument<Int>("adSize") ?: 0
                val listenerId = call.argument<String>("listenerId") ?: ""
                loadBannerAd(slotId, adSize, listenerId, result)
            }
            "showBannerAd" -> {
                val listenerId = call.argument<String>("listenerId") ?: ""
                showBannerAd(listenerId, result)
            }
            "hideBannerAd" -> {
                val listenerId = call.argument<String>("listenerId") ?: ""
                hideBannerAd(listenerId, result)
            }
            "disposeBannerAd" -> {
                val listenerId = call.argument<String>("listenerId") ?: ""
                disposeBannerAd(listenerId, result)
            }
            "loadNativeAd" -> {
                val slotId = call.argument<String>("slotId") ?: ""
                val listenerId = call.argument<String>("listenerId") ?: ""
                @Suppress("UNCHECKED_CAST")
                val style = call.argument<Map<String, Any>>("style")
                loadNativeAd(slotId, listenerId, style, result)
            }
            "disposeNativeAd" -> {
                val listenerId = call.argument<String>("listenerId") ?: ""
                disposeNativeAd(listenerId, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initialize(appId: String, debug: Boolean, result: MethodChannel.Result) {
        // Note: PAGConfig.Builder.setGDPRConsent has been removed since Pangle SDK 7.9.0.9.
        val config = PAGConfig.Builder()
            .appId(appId)
            .debugLog(debug)
            .setPAConsent(1)
            .build()
        
        PAGSdk.init(context, config, object : PAGSdk.PAGInitCallback {
            override fun success() {
                activity?.runOnUiThread {
                    result.success(true)
                }
            }

            override fun fail(code: Int, msg: String?) {
                activity?.runOnUiThread {
                    result.success(false)
                }
            }
        })
    }

    private fun loadSplashAd(slotId: String, timeout: Int, result: MethodChannel.Result) {
        activity?.let { act ->
            splashAdManager = SplashAdManager(act, slotId, timeout, splashEventSink)
            splashAdManager?.loadAd()
            result.success(null)
        } ?: result.error("NO_ACTIVITY", "Activity is null", null)
    }

    private fun loadInterstitialAd(slotId: String, result: MethodChannel.Result) {
        activity?.let { act ->
            interstitialAdManager = InterstitialAdManager(act, slotId, interstitialEventSink)
            interstitialAdManager?.loadAd()
            result.success(true)
        } ?: result.success(false)
    }

    private fun showInterstitialAd(result: MethodChannel.Result) {
        interstitialAdManager?.showAd()
        result.success(null)
    }

    private fun loadBannerAd(slotId: String, adSize: Int, listenerId: String, result: MethodChannel.Result) {
        // Dispose existing banner with same listenerId
        bannerAdManagers[listenerId]?.dispose()
        
        val bannerAdManager = BannerAdManager(context, activity, listenerId, slotId, adSize, channel)
        bannerAdManagers[listenerId] = bannerAdManager
        bannerAdManager.loadAd()
        
        result.success(true)
    }
    
    private fun showBannerAd(listenerId: String, result: MethodChannel.Result) {
        val bannerAdManager = bannerAdManagers[listenerId]
        if (bannerAdManager == null) {
            result.error("NO_BANNER", "Banner ad not found", null)
            return
        }
        
        val success = bannerAdManager.showAd()
        result.success(success)
    }
    
    private fun hideBannerAd(listenerId: String, result: MethodChannel.Result) {
        val bannerAdManager = bannerAdManagers[listenerId]
        if (bannerAdManager == null) {
            result.error("NO_BANNER", "Banner ad not found", null)
            return
        }
        
        bannerAdManager.hideAd()
        result.success(null)
    }
    
    private fun disposeBannerAd(listenerId: String, result: MethodChannel.Result) {
        val bannerAdManager = bannerAdManagers[listenerId]
        bannerAdManager?.dispose()
        bannerAdManagers.remove(listenerId)
        result.success(null)
    }
    
    private fun loadNativeAd(slotId: String, listenerId: String, style: Map<String, Any>?, result: MethodChannel.Result) {
        // Dispose existing manager with the same listenerId (if any)
        nativeAdManagers[listenerId]?.dispose()
        val manager = NativeAdManager(context, activity, listenerId, slotId, style, nativeEventSink)
        nativeAdManagers[listenerId] = manager
        manager.loadAd()
        result.success(true)
    }
    
    private fun disposeNativeAd(listenerId: String, result: MethodChannel.Result) {
        nativeAdManagers[listenerId]?.dispose()
        nativeAdManagers.remove(listenerId)
        result.success(null)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        splashEventChannel.setStreamHandler(null)
        interstitialEventChannel.setStreamHandler(null)
        nativeEventChannel.setStreamHandler(null)
    }
}

File: ios/Classes/pangle_global/PangleAdsHandler.swift
import Flutter
import UIKit
import PAGAdSDK

public class PangleAdsHandler: NSObject {
    var splashEventSink: FlutterEventSink?
    var interstitialEventSink: FlutterEventSink?
    var nativeEventSink: FlutterEventSink?
    private var splashAdManager: SplashAdManager?
    private var interstitialAdManager: InterstitialAdManager?
    private var bannerAdManagers: [String: BannerAdManager] = [:]
    private var nativeAdManagers: [String: NativeAdManager] = [:]
    private var channel: FlutterMethodChannel?
    
    public func register(with registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(name: "multi_ads/pangle_global", binaryMessenger: registrar.messenger())
        channel?.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }
        
        let splashEventChannel = FlutterEventChannel(name: "multi_ads/pangle_global/splash_events", binaryMessenger: registrar.messenger())
        splashEventChannel.setStreamHandler(SplashEventStreamHandler(handler: self))
        
        let interstitialEventChannel = FlutterEventChannel(name: "multi_ads/pangle_global/interstitial_events", binaryMessenger: registrar.messenger())
        interstitialEventChannel.setStreamHandler(InterstitialEventStreamHandler(handler: self))
        
        let nativeEventChannel = FlutterEventChannel(name: "multi_ads/pangle_global/native_events", binaryMessenger: registrar.messenger())
        nativeEventChannel.setStreamHandler(NativeEventStreamHandler(handler: self))
        
        registrar.register(BannerAdViewFactory(messenger: registrar.messenger()), withId: "multi_ads/pangle_global/banner")
        
        // Register banner container for displaying pre-loaded banners
        registrar.register(BannerAdContainerFactory(messenger: registrar.messenger(), handler: self), withId: "multi_ads/pangle_global/banner_container")
        
        // Register native ad view factory
        registrar.register(NativeAdViewFactory(handler: self), withId: "multi_ads/pangle_global/native")
    }
    
    func getBannerView(listenerId: String) -> UIView? {
        return bannerAdManagers[listenerId]?.getBannerView()
    }
    
    func getNativeAd(listenerId: String) -> PAGLNativeAd? {
        return nativeAdManagers[listenerId]?.getNativeAd()
    }
    
    func updateNativeEventSink(_ sink: FlutterEventSink?) {
        self.nativeEventSink = sink
        for manager in nativeAdManagers.values {
            manager.updateEventSink(sink)
        }
    }

    func sendNativeSwipe(listenerId: String, direction: Int) {
        DispatchQueue.main.async {
            self.nativeEventSink?([
                "listenerId": listenerId,
                "event": "onAdSwipe",
                "direction": direction
            ])
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any],
                  let appId = args["appId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            let debug = args["debug"] as? Bool ?? false
            initialize(appId: appId, debug: debug, result: result)
            
        case "loadSplashAd":
            guard let args = call.arguments as? [String: Any],
                  let slotId = args["slotId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            let timeout = args["timeout"] as? Int ?? 3000
            loadSplashAd(slotId: slotId, timeout: timeout, result: result)
            
        case "loadInterstitialAd":
            guard let args = call.arguments as? [String: Any],
                  let slotId = args["slotId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            loadInterstitialAd(slotId: slotId, result: result)
            
        case "showInterstitialAd":
            showInterstitialAd(result: result)
            
        case "loadBannerAd":
            guard let args = call.arguments as? [String: Any],
                  let slotId = args["slotId"] as? String,
                  let adSizeIndex = args["adSize"] as? Int,
                  let listenerId = args["listenerId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            loadBannerAd(slotId: slotId, adSizeIndex: adSizeIndex, listenerId: listenerId, result: result)
            
        case "showBannerAd":
            guard let args = call.arguments as? [String: Any],
                  let listenerId = args["listenerId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            showBannerAd(listenerId: listenerId, result: result)
            
        case "hideBannerAd":
            guard let args = call.arguments as? [String: Any],
                  let listenerId = args["listenerId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            hideBannerAd(listenerId: listenerId, result: result)
            
        case "disposeBannerAd":
            guard let args = call.arguments as? [String: Any],
                  let listenerId = args["listenerId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            disposeBannerAd(listenerId: listenerId, result: result)
            
        case "loadNativeAd":
            guard let args = call.arguments as? [String: Any],
                  let slotId = args["slotId"] as? String,
                  let listenerId = args["listenerId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            loadNativeAd(slotId: slotId, listenerId: listenerId, result: result)
            
        case "disposeNativeAd":
            guard let args = call.arguments as? [String: Any],
                  let listenerId = args["listenerId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            disposeNativeAd(listenerId: listenerId, result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initialize(appId: String, debug: Bool, result: @escaping FlutterResult) {
        let config = PAGConfig.share()
        config.appID = appId
        config.debugLog = debug
        
        config.paConsent = .consent
        config.gdprConsent = .consent
        
        PAGSdk.start(with: config) { success, error in
            if success {
                result(true)
            } else {
                result(false)
            }
        }
    }
    
    private func loadSplashAd(slotId: String, timeout: Int, result: @escaping FlutterResult) {
        splashAdManager = SplashAdManager(slotId: slotId, timeout: timeout, eventSink: splashEventSink)
        splashAdManager?.loadAd()
        result(nil)
    }
    
    private func loadInterstitialAd(slotId: String, result: @escaping FlutterResult) {
        interstitialAdManager = InterstitialAdManager(slotId: slotId, eventSink: interstitialEventSink)
        interstitialAdManager?.loadAd()
        result(true)
    }
    
    private func showInterstitialAd(result: @escaping FlutterResult) {
        interstitialAdManager?.showAd()
        result(nil)
    }
    
    private func loadBannerAd(slotId: String, adSizeIndex: Int, listenerId: String, result: @escaping FlutterResult) {
        guard let channel = self.channel else {
            result(FlutterError(code: "NO_CHANNEL", message: "Method channel not available", details: nil))
            return
        }
        
        // Convert adSizeIndex to PAGBannerAdSize
        let bannerSize: PAGBannerAdSize
        switch adSizeIndex {
        case 0:
            bannerSize = kPAGBannerSize320x50
        case 1:
            bannerSize = kPAGBannerSize300x250
        case 2:
            bannerSize = kPAGBannerSize728x90
        case 3:
            let screenWidth = UIScreen.main.bounds.width
            bannerSize = PAGCurrentOrientationAnchoredAdaptiveBannerAdSizeWithWidth(screenWidth)
        default:
            bannerSize = kPAGBannerSize320x50
        }
        
        // Dispose existing banner with same listenerId
        if let existingManager = bannerAdManagers[listenerId] {
            existingManager.dispose()
        }
        
        let bannerAdManager = BannerAdManager(listenerId: listenerId, slotId: slotId, adSize: bannerSize, channel: channel)
        bannerAdManagers[listenerId] = bannerAdManager
        bannerAdManager.loadAd()
        
        result(true)
    }
    
    private func showBannerAd(listenerId: String, result: @escaping FlutterResult) {
        guard let bannerAdManager = bannerAdManagers[listenerId] else {
            result(FlutterError(code: "NO_BANNER", message: "Banner ad not found", details: nil))
            return
        }
        
        let success = bannerAdManager.showAd()
        result(success)
    }
    
    private func hideBannerAd(listenerId: String, result: @escaping FlutterResult) {
        guard let bannerAdManager = bannerAdManagers[listenerId] else {
            result(FlutterError(code: "NO_BANNER", message: "Banner ad not found", details: nil))
            return
        }
        
        bannerAdManager.hideAd()
        result(nil)
    }
    
    private func disposeBannerAd(listenerId: String, result: @escaping FlutterResult) {
        guard let bannerAdManager = bannerAdManagers[listenerId] else {
            result(nil)
            return
        }
        
        bannerAdManager.dispose()
        bannerAdManagers.removeValue(forKey: listenerId)
        result(nil)
    }
    
    private func loadNativeAd(slotId: String, listenerId: String, result: @escaping FlutterResult) {
        // Dispose existing manager with same listenerId
        if let existing = nativeAdManagers[listenerId] {
            existing.dispose()
        }
        let manager = NativeAdManager(listenerId: listenerId, slotId: slotId, eventSink: nativeEventSink)
        nativeAdManagers[listenerId] = manager
        manager.loadAd()
        result(true)
    }
    
    private func disposeNativeAd(listenerId: String, result: @escaping FlutterResult) {
        if let manager = nativeAdManagers[listenerId] {
            manager.dispose()
            nativeAdManagers.removeValue(forKey: listenerId)
        }
        result(nil)
    }
}

class SplashEventStreamHandler: NSObject, FlutterStreamHandler {
    weak var handler: PangleAdsHandler?
    
    init(handler: PangleAdsHandler) {
        self.handler = handler
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        handler?.splashEventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        handler?.splashEventSink = nil
        return nil
    }
}

class InterstitialEventStreamHandler: NSObject, FlutterStreamHandler {
    weak var handler: PangleAdsHandler?
    
    init(handler: PangleAdsHandler) {
        self.handler = handler
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        handler?.interstitialEventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        handler?.interstitialEventSink = nil
        return nil
    }
}

class NativeEventStreamHandler: NSObject, FlutterStreamHandler {
    weak var handler: PangleAdsHandler?
    
    init(handler: PangleAdsHandler) {
        self.handler = handler
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        handler?.updateNativeEventSink(events)
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        handler?.updateNativeEventSink(nil)
        return nil
    }
}
